model-optimizer/extensions/middle/TFDepthwiseConv2dNativeReshape.py

```diff
@@ -0,0 +1,41 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+from extensions.middle.ConvertLayoutDependentOperations import ConvertLayoutDependentOperations
+from mo.front.common.partial_infer.utils import int64_array
+from mo.front.tf.graph_utils import create_op_node_with_second_input
+from mo.graph.graph import Graph
+from mo.middle.replacement import MiddleReplacementPattern
+from mo.ops.reshape import Reshape
+
+
+class TFDepthwiseConv2dNativeReshape(MiddleReplacementPattern):
+    """
+    According to the specification of TF* DepthwiseConv2dNative kernel has shape
+    [filter_height, filter_width, in_channels, channel_multiplier], and output_channels = in_channels * channel_multiplier.
+    In general, we convert this operation to a GroupConvolution, by setting group = in_channels. So in IR kernel shape will be
+    [in_channels, channel_multiplier, 1, filter_height, filter_width] <-> [G, O, I, H, W] (extensions/back/ConvolutionNormalizer.py)
+
+    But there is a case when in_channels = 1 and it is converted to a Convolution, so we can not calculate the number of
+    output_channels correctly. We will transpose weights from [filter_height, filter_width, in_channels, channel_multiplier]
+    to [in_channels, channel_multiplier, filter_height, filter_width]. This is incorrect, because for Convolution op
+    kernel shape should be [in_channels * channel_multiplier, 1, filter_height, filter_width] according to OpenVINO
+    specification <-> [O, I, H, W].
+
+    So we have to reshape kernel shape to [filter_height, filter_width, in_channels * channel_multiplier, 1] to transpose
+    it correctly later
+    """
+    def run_before(self):
+        return [ConvertLayoutDependentOperations]
+
+    enabled = True
+    force_clean_up = True
+
+    def find_and_replace_pattern(self, graph: Graph):
```

Comment:
I don't like the idea of this additional transformation. Can we modify the transformation in `extensions/back/ConvolutionNormalizer.py` to convert this `DepthwiseConv2dNative` operation to a GroupConvolution as it should be (instead of Convolution which it is converted now)? In this case we will have all Convolution normalization logic in one place.

@yekruglov , @rkazants , what do you think about this?

Reply:
@lazarevevgeny, I don't like this solution either, unfortunately.
Per my understanding `DepthwiseConv2d` is not supported by MO tool at all despite we have extractor for this operation but it seems to have incorrect implementation. And it needs total re-implementation activity. Now we have a mess for `DepthwiseConv` with `Conv` and `GroupConv` in the MO tool so that at the back stage we don't know that this operation has been `DepthwiseConv` originally.
For correct implementation and `GroupConv` utilize we need to modify original kernel by repeating it `in_channels` times along channel dimension in order to provide split data into `in_channels` groups and apply `depthwisemultiplier` depthwise filters per group.
We can implement internal MO operation `DepthwiseConv` to save all parameters and on the middle stage to convert it into `GroupConvolution`.

Reply:
I actually don't understand this statement: 

> But there is a case when in_channels = 1 and it is converted to a Convolution, so we can not calculate the number of    output_channels correctly.

Why we can't alway use GroupConvolution even if in_channels = 1? If we have GroupConvolution to Convolution converter and it works wrong, then it should be fixed. Othrwsie I don't see any reason not to use GroupConvolution for `DepthwiseConv2dNative`.

Reply:
> Why we can't alway use GroupConvolution even if in_channels = 1? If we have GroupConvolution to Convolution converter and it works wrong, then it should be fixed.

This is exactly my point. The transformation in the `extensions/back/ConvolutionNormalizer.py` which creates GroupConvolution (which currently doesn't work for the case when `in_channels = 1`) should be modified to support case when `in_channels = 1`. And we can do this by checking the value of `op` attribute of the node which contains value `DepthwiseConv2d`. 

@GlebKazantaev is that you mean?

Reply:
@lazarevevgeny, @GlebKazantaev, I think MO doesn't work correctly for `DepthwiseConv` not only `in_channels=1` case. Other cases `in_channels>1` are not supported either since I haven't seen filter kernel transformation to make it applicable for `GroupConv` operation.

Reply:
I also think that we should add layer tests for the TF `DepthwiseConv`to make sure that we convert the models correctly.